### PR TITLE
Move Snyk reports from GitHub Security tab to GitHub issues

### LINKED ISSUE #29494: Move Snyk reports from GitHub Security tab to GitHub issues
### Problem Statement

Currently, security alerts for publicly known CVEs in third-party libraries are uploaded as a SARIF file under the Security tab on GitHub. While this is a positive step, it does not provide sufficient visibility or raise awareness within our team. The process requires team members to visit the Security tab daily, teams are unable to add the task into their backlog, and despite the security scanner being fully functional, there is a risk of missing critical CVE alerts.

### Proposal

To enhance visibility and ensure our team is aware of CVEs on third-party libraries, I would like to propose the following:

1. **Report migration**  
   Develop a script or use an existing tool to automate the creation of CVEs reported by Snyk to GitHub issues. This will make CVE alerts more accessible and actionable for the team. 

2. **Label assignment**  
   Assign the following labels to each created issue to prioritize them effectively:
   - `status/triage`
   - `kind/cve`
   - `status/important`

3. **Centralize**  
   Discontinue the use of the GitHub Security tab as the primary source for CVE alerts. Instead, use GitHub issues as the central repository for tracking and managing security vulnerabilities on third-party libraries.

By implementing this enhancement, we aim to improve the visibility of security alerts, ensure timely response to CVEs, and streamline our CVE management processes.

### PR DESCRIPTION
**Pull request by abstractj**
Closes #29494

For more details, please check the issue description. Also, [I have a running PoC here](https://github.com/keycloak-poc/keycloak).

### CODE REVIEW

File: .github/workflows/snyk-analysis.yml [hunk 1]
```diff
@@ -24,28 +24,17 @@ jobs:
       - uses: snyk/actions/setup@master
 
       - name: Check for vulnerabilities in Quarkus
-        run: snyk test --policy-path=${GITHUB_WORKSPACE}/.github/snyk/.snyk --all-projects --prune-repeated-subdependencies --exclude=tests --sarif-file-output=quarkus-report.sarif quarkus/deployment
+        run: snyk test --policy-path=${GITHUB_WORKSPACE}/.github/snyk/.snyk --all-projects --prune-repeated-subdependencies --exclude=tests --json quarkus/deployment | .github/scripts/snyk-report.sh "$GITHUB_TOKEN"
```
Comment (jonkoops) [inline .github/workflows/snyk-analysis.yml:27]: Why pass the GitHub token as an argument if it is already being passed as an environment variable?

File: .github/scripts/snyk-report.sh [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+# Prevent duplicates by checking if a similar CVE exists
+check_github_issue_exists() {
+    local issue_title="$1"
+    # Extract the string before the last hyphen
+    local extracted_title="${issue_title%-*}"
+    
+    # Encode the extracted title for the URL
+    local encoded_title=$(echo "$extracted_title" | jq -sRr @uri)
+    local search_url="https://api.github.com/search/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+$encoded_title+repo:$KEYCLOAK_REPO"
```
Comment (rmartinc) [inline .github/scripts/snyk-report.sh:13]: Is it Ok if the issue exists but closed? What happens if we decide to no fix it? aybe we should add something for those cases. I don't want to have issues re-openen all teh time. :smile:

**Review by rmartinc — CHANGES_REQUESTED** [review]
@abstractj Just one question about issues that we decided to not fix (for any reason). Maybe we need something special for them (a label or something we can discard them as already evaluated).

**Comment by abstractj** [comment]
@jonkoops @rmartinc thanks for reviewing the changes. I've incorporated your suggestions and fixed some minor bugs I discovered along the way. Here's a summary of the updates:

- Use the SNYK ID in scenarios where there's no CVE ID to prevent duplicates.
- Removed GITHUB_TOKEN as a parameter from the script.
- Added checks for both open and closed issues.

[I ran some tests](https://github.com/keycloak-poc/keycloak/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc), and everything appears to be working well.

@stianst @ahus1 @vmuzikar @mposolda considering that our goal is to increase awareness in our team's bug dashboards, could you please confirm if the labels "status/triage," "kind/cve," and "status/important" are sufficient from your perspective? Should we add more labels?

**Review by jonkoops — DISMISSED** [review]

**Comment by abstractj** [comment]
> @abstractj Just one question about issues that we decided to not fix (for any reason). Maybe we need something special for them (a label or something we can discard them as already evaluated). 

@rmartinc for that we will close the issue and mark as "Not planned".

**Review by ahus1 — DISMISSED** [review]

**Comment by stianst** [comment]
> > @abstractj Just one question about issues that we decided to not fix (for any reason). Maybe we need something special for them (a label or something we can discard them as already evaluated).
> 
> @rmartinc for that we will close the issue and mark as "Not planned".

Probably with an explanation right?

File: .github/scripts/snyk-report.sh [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+# Create a GH issue based on the content of the CVE
+create_github_issue() {
+    local title="$1"
+    local body="$2"
+
+    local api_url="https://api.github.com/repos/$KEYCLOAK_REPO/issues"
+    local data=$(jq -n --arg title "$title" --arg body "$body" \
+                 '{title: $title, body: $body, labels: ["status/triage", "kind/cve", "status/important"]}')
```
Comment (stianst) [inline .github/scripts/snyk-report.sh:28]: We don't have a label `status/important`, we have priority labels, but those should be set as part of triage, not automatically.

**Review by stianst — CHANGES_REQUESTED** [review]
Creating as issues is one thing, but what will the process be for triage these issues?

I would suggest let's also add `kind/bug` label, and just triage these as part of regular bug triaging process. One question there though is how/who do we route these to the correct team (setting team labels).

**Comment by abstractj** [comment]
> Probably with an explanation right?

Always, that's accurate, and we can even add that as part of the bug triage documentation.

**Comment by abstractj** [comment]
> Creating as issues is one thing, but what will the process be for triage these issues?
> 
> I would suggest let's also add `kind/bug` label, and just triage these as part of regular bug triaging process. One question there though is how/who do we route these to the correct team (setting team labels).

Thank you! I will make the changes regarding this label. As for the "how" and "who," I will continue to review them on the daily basis with @jonkoops and @rmartinc. Additionally, we have the "Missing team" report in the bugs' dashboard, so we can make it part of the status for our weekly meetings. Does that make sense?

File: .github/scripts/snyk-report.sh (hunk 3 above)
Comment (abstractj) [reply .github/scripts/snyk-report.sh]: @stianst updated

**Review by jonkoops — APPROVED** [review]

**Review by rmartinc — APPROVED** [review]
LGTM!

**Review by vmuzikar — COMMENTED** [review]
@abstractj Thanks for this enhancement, nice stuff.

I share the same concern as Stian – how to automate routing to the correct team. I wonder if it would be very distracting to tag all sub-teams on each vulnerability discovered by the automated scanners, so that a correct team self-assigns it. That would help reducing the bus factor of only a few people doing the triage.

On top of that, I was wondering if we really want to get rid of the GH Security tab. It provides a nice UI and the ability to link GH Issue, see this [example](https://github.com/vmuzikar/keycloak/security/code-scanning/260). Should we rather keep the Security and just link the Issues to it?

**Comment by stianst** [comment]
I think this works as a start; and someone from sec team or me can assign to individual teams for now. Would be good in the future to do this automatically though, and we can probably relatively easily do a list of what dependencies each team owns or something like that.

**Review by stianst — APPROVED** [review]
LGTM as a start, we can tweak later as needed

**Comment by abstractj** [comment]
> LGTM as a start, we can tweak later as needed

Created this as a follow up https://github.com/keycloak/keycloak/issues/29617

**Pull request merged**